integrated_channels/canvas/client.py

```diff
@@ -378,37 +378,56 @@ def _handle_canvas_assignment_retrieval(
         canvas_assignments_url = '{canvas_base_url}/api/v1/courses/{course_id}/assignments'.format(
             canvas_base_url=self.enterprise_configuration.canvas_base_url,
             course_id=course_id
         )
         resp = self.session.get(canvas_assignments_url)
 
-        if resp.status_code >= 400:
-            raise ClientError(
-                "Something went wrong retrieving assignments from Canvas. Got response: {}".format(
-                    resp.text,
-                ),
-                resp.status_code
-            )
+        more_pages_present = True
+        assignment_id = ''
 
-        assignments_resp = resp.json()
-        assignment_id = None
-        for assignment in assignments_resp:
-            try:
-                if assignment['integration_id'] == integration_id:
-                    assignment_id = assignment['id']
-                    break
-            # The validation check above should ensure that we have a 200 response from Canvas, but sanity catch if we
-            # have a unexpected response format
-            except (KeyError, ValueError, TypeError) as error:
+        while more_pages_present:
+            if resp.status_code >= 400:
                 raise ClientError(
                     "Something went wrong retrieving assignments from Canvas. Got response: {}".format(
                         resp.text,
                     ),
                     resp.status_code
-                ) from error
+                )
+
+            assignments_resp = resp.json()
+            for assignment in assignments_resp:
+                try:
+                    if assignment['integration_id'] == integration_id:
+                        assignment_id = assignment['id']
+                        break
+
+                # The validation check above should ensure that we have a 200 response from Canvas,
+                # but sanity catch if we have a unexpected response format
+                except (KeyError, ValueError, TypeError) as error:
+                    raise ClientError(
+                        "Something went wrong retrieving assignments from Canvas. Got response: {}".format(
+                            resp.text,
+                        ),
+                        resp.status_code
+                    ) from error
+
+            if not assignment_id:
+                # Canvas pagination headers come back as a string,
+                # so we have to parse out the linked list of assignment pages
+                assignment_page_results = resp.headers['Link'].split(',')
+                current_page = assignment_page_results[0].split(';')[0].strip('<>')
+                last_page = assignment_page_results[3].split(';')[0].strip('<>')
+
+                if current_page == last_page:
+                    more_pages_present = False
+                else:
+                    next_page_url = assignment_page_results[1].split(';')[0].strip('<>')
```

Comment:
Nit: I don't love the loop style here where you have 1 get request outside the loop and 1 inside it. It's harder to reason about the code when I come back to read it. Next time consider a loop like this:
more_pages_present = True
while more_pages_present: 
   url = some_url_string_logic()
   resp = self.session(url)
   done = do_stuff_with response()
   if done: break

Reply:
Agreed, I shot myself in the foot by trying to preserve the original beginning of the function. If you feel as though it's too unreadable I could be convinced to rearrange things but do at least believe that this should function as is.